Implement blocking Dequeue and Peek

Kriechi opened the pull request:
fixes #20 

@joncrlsn do you prefer code duplication for readability, or extracting a "blocking" function that takes a `func` as argument?

Kriechi committed:
implement blocking Dequeue and Peek functions

This uses a sync.Cond object to sleep multiple goroutines and wake them
as soon as a new item is enqueued.

joncrlsn commented on queue.go:
```diff
@@ -175,6 +181,9 @@ func (q *DQue) Enqueue(obj interface{}) error {
 	}
 
+	// Wakeup any goroutine that is currently waiting for an item to be enqueued
+	q.emptyCond.Broadcast()
```
Is Broadcast() identical to Signal()?  It seems like it is because there can be only one emptyCond.Wait() happening at a time (given the lock surrounding it).  Just asking.

Kriechi replied:
Broadcast wakes up ALL waiting goroutines.
Signal only one of them.

The Wait() call atomically unlocks & sleeps. I added a more descriptive comment to the code.

joncrlsn commented on queue.go:
```diff
@@ -260,6 +269,40 @@ func (q *DQue) Peek() (interface{}, error) {
+		if err == ErrEmpty {
+			q.mutexEmptyCond.Lock()
+			q.emptyCond.Wait()
+			q.mutexEmptyCond.Unlock()
```
It feels like there is room for a race condition here...  where you could get an ErrEmpty, but something is placed into the queue and the Broadcast() call is made before you hit the Wait().  So then you'd be waiting when there was something in the queue.

Kriechi replied:
You are right - I pushed a fix (more like a rewrite).

Kriechi commented on example_test.go:
```diff
@@ -7,8 +7,9 @@ package dque_test
-	"github.com/joncrlsn/dque"
 	"log"
+
+	"github.com/joncrlsn/dque"
```
That was my VSCode editor (and one of the go-fmt / linters). Objections?

joncrlsn replied:
No :-)

Kriechi commented on segment.go:
```diff
@@ -84,12 +84,12 @@ func (seg *qSegment) load() error {
-		seg.file = f
 	}
+	defer f.Close()
+	seg.file = f
```
My linter / editor complained about a useless else block with a prior return statement.

joncrlsn reviewed (commented):
Hi Thomas, This looks good!  I didn't realize that a whole week had gone by since you submitted your fix.  I think this will work, but what do you think about having a more aggressive test?  Like having an enqueueing goroutine that does 3 enqueues separated by about 5-50ms (or whatever).  Then have three dequeue goroutines that each expect to dequeue one item, and then a final dequeue that expects to get nothing.  You've kind of done that, but not with multiple goroutines waiting on the same queue.

Or, when I've done things like this in the past I've tried to add some randomness to the amount of time that it sleeps (0ms to 100ms or less) before enqueueing and then have it loop 100 times or so just to make sure there aren't any hidden race conditions that only show up under load.  That was with more manual testing where I could kill the process after a certain period of time, but maybe you could do something similar in a go test if you had each dequeing goroutine have a shared count of dequeues so they would know to die once the number of dequeues had been reached?  I haven't done a lot of goroutines so I don't remember in what circumstances they will keep the testing process alive.

Thoughts?

Kriechi committed:
blocking: add more aggressive test

Kriechi commented:
I added a new test and ran `go test -run TestQueue_BlockingAggresive -count 100 -failfast` which  succeeded without issues after 441 seconds.

I also added a small fix + test for `Close` while waiting in a blocking call.

Kriechi committed:
fix data race - always lock the queue before doing anything

Kriechi commented on queue.go:
```diff
@@ -178,14 +191,14 @@ func (q *DQue) Close() error {
+	if q.fileLock == nil {
+		return ErrQueueClosed
+	}
+
```
I had to swap these two instruction blocks:
Old:
* check if queue is closed
* potentially somebody else calls close
* get the mutex lock
* access potentially closed queue

New:
* get the mutex lock
* nobody else can close this queue while we have the lock
* check if closed
* access a 100% open queue
